### PR TITLE
Simplify e2e.RunRC method and wait up to 10 minutes for pods to start

### PR DESCRIPTION
**Pull request by fgrzadkowski**
This simplifies e2e.RunRC which was extremely long. Now, that we use watch-based cache we can query for all pods as often as we want to count create/running pods. It also increases timeout for waiting for pods to start to 10 minutes to address flakiness in load test.

@rrati AFAIU this PR keeps all the requirements, i.e. query all pods every second and count create and running pods.

### CODE REVIEW

**Comment by fgrzadkowski** [comment]
Fixes #7561

File: test/e2e/util.go [hunk 1]
```diff
@@ -856,179 +841,62 @@ func RunRC(config RCConfig) error {
...
-				return fmt.Errorf("Controller %s: Number of submitted pods dropped from %d to %d", name, last, current)
-			}
-
-			if same >= failCount {
-				return fmt.Errorf("Controller %s: No pods submitted for the last %d checks", name, failCount)
-			}
-			last = current
-		}
+	interval := config.PollInterval
+	if interval < 0 {
```
Comment (gmarek) [inline test/e2e/util.go]: < or <=? I.e. not set == 0 and I guess you wanted to set the 'default' value here.

File: test/e2e/util.go [hunk 2]
```diff
@@ -856,179 +841,62 @@ func RunRC(config RCConfig) error {
...
-						pending++
-					}
-				} else if p.Status.Phase == api.PodSucceeded || p.Status.Phase == api.PodFailed {
-					inactive++
-				} else if p.Status.Phase == api.PodUnknown {
-					unknown++
+	start := time.Now()
+	running := 0
+	for running < config.Replicas && time.Since(start) < 10*time.Minute {
+		time.Sleep(interval)
```
Comment (rrati) [inline test/e2e/util.go]: This isn't guaranteed to scale.  30 pods on 10 nodes might start in 10 minutes, but can you guarantee 100 pods on 100 nodes would?  The old looping mechanism didn't care about how long it took, only that there was progress.  Having a hard time limit makes this function less reliable for different pod counts.  I'd prefer going back to tracking progress vs arbitrary timeout

**Comment by gmarek** [comment]
LGTM, but it drops some unimportant checks. As this PR slightly changes semantics of RunRC I'd like to have someone else look at it.

File: test/e2e/util.go (hunk 2 above)
Comment (gmarek) [inline test/e2e/util.go]: I don't think we should worry about 10000 pods yet.

File: test/e2e/util.go [hunk 3]
```diff
@@ -856,179 +841,62 @@ func RunRC(config RCConfig) error {
...
-					} else {
-						pending++
-					}
-				} else if p.Status.Phase == api.PodSucceeded || p.Status.Phase == api.PodFailed {
-					inactive++
-				} else if p.Status.Phase == api.PodUnknown {
-					unknown++
+	start := time.Now()
+	running := 0
+	for running < config.Replicas && time.Since(start) < 10*time.Minute {
```
Comment (gmarek) [inline test/e2e/util.go]: running != config.Replicas

File: test/e2e/util.go (hunk 2 above)
Comment (rrati) [inline test/e2e/util.go]: There's no reason to not allow for it. We run 100 pods/node on 23 physical nodes (2300 pods) and with performance regressions we've seen, only see about 3 pods start/sec (if the math in #8119 is correct).  That would be roughly 700 seconds, or >10 minutes, for all pods to start.  A hard time limit isn't the right path here.

Comment (gmarek) [inline test/e2e/util.go]: We're currently targeting 30 pods/node.

Comment (rrati) [inline test/e2e/util.go]: Again, no reason to prevent running at larger scale.

File: test/e2e/util.go [hunk 4]
```diff
@@ -856,179 +841,62 @@ func RunRC(config RCConfig) error {
...
 			}
+		}
 
-			Logf("Pod States: %d running, %d pending, %d waiting, %d inactive, %d unknown ", current, pending, waiting, inactive, unknown)
-			if config.PodStatusFile != nil {
-				fmt.Fprintf(config.PodStatusFile, "%s, %d, running, %d, pending, %d, waiting, %d, inactive, %d, unknown\n", item.createTime, current, pending, waiting, inactive, unknown)
-			}
-
-			if foundAllPods && len(currentPods) != len(oldPods) {
-
```
Comment (rrati) [inline test/e2e/util.go]: Why are we removing this condition?  Is this a state we don't want to catch anymore?  There haven been issues with pod start up in the past.  Why wouldn't we want to continue looking for them?

File: test/e2e/util.go [hunk 5]
```diff
@@ -856,179 +841,62 @@ func RunRC(config RCConfig) error {
...
-				errorStr := fmt.Sprintf("Number of reported pods changed: %d vs %d", len(currentPods), len(oldPods))
-				Logf("%v, pods that changed since the last iteration:", errorStr)
-				Diff(oldPods, currentPods).Print(util.NewStringSet())
-				return fmt.Errorf(errorStr)
-			}
-			if last < current {
-				same = 0
-			} else if last == current {
-				same++
-			} else if current < last {
```
Comment (rrati) [inline test/e2e/util.go]: Same as above

File: test/e2e/util.go [hunk 6]
```diff
@@ -856,179 +841,62 @@ func RunRC(config RCConfig) error {
...
-
-				// The pod failed or succeeded, or was somehow pushed out of running by the kubelet.
-				errorStr := fmt.Sprintf("Number of running pods dropped from %d to %d", last, current)
-				Logf("%v, pods that changed since the last iteration:", errorStr)
-				Diff(oldPods, currentPods).Print(util.NewStringSet())
-				return fmt.Errorf(errorStr)
-			}
-			if same >= failCount {
```
Comment (rrati) [inline test/e2e/util.go]: Same as above

File: test/e2e/util.go (hunk 2 above)
Comment (timothysc) [inline test/e2e/util.go]: +1 to rrati no reason to change that aspect.

**Comment by rrati** [comment]
@fgrzadkowski I mostly agree with this simplification. Moving to the pod store allowed this to be much cleaner.  I think there are some things that need to be addressed/investigated in this changes though, like the hard pod start limit and the removal of the pod startup failure condition checks.

File: test/e2e/util.go (hunk 2 above)
Comment (lavalamp) [inline test/e2e/util.go]: Suggestion for timeout: 30 seconds + 2 seconds per pod?

File: test/e2e/util.go (hunk 5 above)
Comment (fgrzadkowski) [inline test/e2e/util.go]: As you said this is tested by failures.

File: test/e2e/util.go (hunk 4 above)
Comment (fgrzadkowski) [inline test/e2e/util.go]: Done

File: test/e2e/util.go (hunk 2 above)
Comment (fgrzadkowski) [inline test/e2e/util.go]: Done

File: test/e2e/util.go (hunk 3 above)
Comment (fgrzadkowski) [inline test/e2e/util.go]: Done

File: test/e2e/util.go (hunk 1 above)
Comment (fgrzadkowski) [inline test/e2e/util.go]: Done

**Comment by fgrzadkowski** [comment]
@rrati @gmarek I addressed all the comments by re-adding checks and removing hard deadline. PTAL

File: test/e2e/util.go [hunk 7]
```diff
@@ -808,47 +808,32 @@ func Diff(oldPods []*api.Pod, curPods []*api.Pod) PodDiff {
 }
 
 // RunRC Launches (and verifies correctness) of a Replication Controller
-// and will wait for all pods it spawns to become "Running".
+// and will wait all pods it spawns to become "Running".
```
Comment (rrati) [inline test/e2e/util.go]: wait for all ...

**Comment by gmarek** [comment]
LGTM. I'm going to merge on green, so we'll see if Jenkins will be happier with this change. @rrati - if you want some more changes done, please let @fgrzadkowski know.

**Comment by rrati** [comment]
@gmarek I am reviewing now

File: test/e2e/util.go (hunk 7 above)
Comment (fgrzadkowski) [inline test/e2e/util.go]: Done

**Comment by rrati** [comment]
LGTM

**Pull request merged**